### ISSUE TITLE
restic should not read CACHEDIR.TAG in parent directories

### ISSUE THREAD

**Issue by dionorgua** [restic/restic#2722]
Output of `restic version`
--------------------------
restic 0.9.6 compiled with go1.13.4 on linux/amd64
restic 0.9.6 (v0.9.6-200-g39f58a2b-dirty) compiled with go1.14.2 on linux/amd64

How did you run restic exactly?
-------------------------------

```
/tmp/sample_dir% tree
.
├── CACHEDIR.TAG
└── src
    └── backup_me.txt

/tmp/sample_dir% restic backup --exclude-caches $PWD/src

% restic ls latest                       
repository 22fc71ff opened successfully, password is correct
snapshot 3c6a0afd of [/tmp/sample_dir/src] filtered by [] at 2020-05-07 15:54:30.679914516 +0300 EEST):
/tmp
/tmp/sample_dir
```

What backend/server/service did you use to store the repository?
----------------------------------------------------------------
local/rest-server

Expected behavior
-----------------
I think that `tar` behavior should be preserver here
```
/tmp/sample_dir% tar czvf test.tar.gz --exclude-caches $PWD/src
tar: Removing leading `/' from member names
/tmp/sample_dir/src/
/tmp/sample_dir/src/backup_me.txt
```
So if user explicitly asked to backup `/tmp/sample_dir/src` then `restic` should not look for `CACHEDIR.TAG` files in parent directories. 

Actual behavior
---------------
restic doesn't backup files if `CACHEDIR.TAG` is present in parent directory

Steps to reproduce the behavior
-------------------------------
See above

Do you have any idea what may have caused this?
-----------------------------------------------

Do you have an idea how to solve the issue?
-------------------------------------------

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------

Excellent tool

**Comment by rawtaz** [comment]
Well, as you probably understand, the current design is that 1) `CACHEDIR.TAG` signifies that "the dir this file resides in, is of a cache nature, and hence nothing in it should be backed up".

If you want parts of that directory to be backed up, why do you have a `CACHEDIR.TAG` file in it anyway? If you do have contents in it that you want to back up, it's clearly not a cache directory.

**Comment by dionorgua** [comment]
I understand that this is is some sort of corner case. At the same time it may make sense for something like this:
```
├── CACHEDIR.TAG
└── src
    ├── backup_me.txt
    ├── some_dir
    └── src2
        ├── CACHEDIR.TAG
        └── large_file
```
Imagine that I've something called `~/tmp` that I usually use as working directory for something not important. So it's reasonable to have `~/tmp/CACHEDIR.TAG` file there to don't backup it by default.  

Now I've just downloaded something (or restored almost died HDD) to `~/tmp/some_data` and want to backup it before cleanup. 

It's just counter-intuitive to think that backup program may look for files/directories outside of source directory. `tar`, `borg` and other tools are good example for this.

**Comment by rawtaz** [comment]
I'm sorry, but that is too specific of a corner-case to justify changing the behavior of the `CACHEDIR.TAG` feature. Some suggestions:

- If you have only volatile stuff you don't intend to back up in `/tmp`, which is perfectly fine use of that directory, don't include it in the backup set in the first place/at all.

- Also, `/tmp` is just a temporary directory, so you shouldn't expect the `CACHEDIR.TAG` file to persist in there either - on the contrary you should expect it to be erased on every system reboot.

- Put your downloads that you may want to back up somewhere else than in `/tmp`. It doesn't make sense to use `/tmp` for that.

I'm closing this for now as I think the solution to your use case is to change your directory structure or use of them :)

**Comment by dionorgua** [comment]
I've mentioned `~/tmp` (`$HOME/tmp`, not system `/tmp` directory). If you don't like this name, use `/home/user/no_backup_by_default`. 

Also about `CACHEDIR.TAG` name. It's also not important here and has nothing to do with caches. It maybe called like `.nobackup` and `--exclude-if-present .nobackup`. 

As I said, this is common behavior for all tools that supports `exclude-if-present` like idiom. I don't know another tool that has this restic-like behavior. For backup app such unexpected behavior may cause data loss so I strongly suggest to reconsider this.

**Comment by dionorgua** [comment]
Imagine what hell will happens if `git status` will not show anything because you've something in `.gitignore` that is placed in parent repository for multirepo project.

**Comment by rawtaz** [comment]
You are correct, you mentioned `~/tmp` rather than `/tmp`, I should probably have used just `tmp/` instead. Yes, the name of `CACHEDIR.TAG` isn't relevant. The function of it is.

But why do you put files in a directory that you have explicitly designated as a directory that should not be backed up, and then expect those files to be backed up? If you want files to be backed up, put them in another place than a directory that you have explicitly said should not be backed up?

**Comment by dionorgua** [comment]
As I mentioned here I use `CACHEDIR.TAG`-like files to mark some 'working' directories as not backed up by default. Default means automatic daily backup from cron job that may happens at any time and with usually long `keep-*` forget policy. 

So I usually use such directories for any work with a lot of large files with very short lifetime. I even have a quck script called `mark_nobackup` that creates such `CACHEDIR.TAG`-like files.

This time it was HDD restore from another computer and attempt to backup it's content after this.

It is so common for me that I'll need to avoid usage of `--exclude-if-present` at all until this is fixed (still have magic around `find . -name nobackup` commented out in my cron script from times when there was no `exclude-if-present` in restic)

**Comment by dionorgua** [comment]
PS. I can fix this myself and send pull request if it has chances to be merged.

**Comment by rawtaz** [comment]
Again though; why don't you just put your files that you *want* backed up in another directory than where you have the `CACHEDIR.TAG` file? It's just one directory away.

The problem is that the current behavior is what it is. If we change that, many many other users may see unexpected results. I don't envision many users asking to back up specific dirs/files under a directory that has a `CACHEDIR.TAG` file, but we don't know, and in particular your use case for this is quite specific.

**Comment by dionorgua** [comment]
1. I have periodic cron job that performs daily backup of my whole system (like `restic backup --exclude-caches /`). This job is called from `/etc/cron.daily` so may happens at any time during day.
 
2. I need to temporary copy HDD from another machine. So I create new directory, mark it as `nobackup` to avoid periodic daily backup. If I don't do this, it may happens that daily backup will be started and restic will backup this dir and this will take a lot of storage. It also may happens that such unexpected daily backup will be also kept as 'weekly/monthly/yearly' snapshot and will consume a lot of storage for a long time.

3. After copying files from HDD I want to backup this directory _manually_ and alone. This snapshot will be properly tagged and may be removed later at any time. It'll be easy to find it in `restic snapshots` output and it'll be my decision whether to keep it or not and when to remove. 

I may agree that for `--exclude-caches` case it makes at least some sense. But definitely not for generic  `--exclude-if-present` case. I've just googled for `--exclude-if-present` and checked followed programs:
* GNU tar (--exclude-caches)
* borg (--exclude-caches)
* rclone (--exclude-if-present)
* duplicity (--exclude-if-present)
* ridff_backup (--exclude-if-present)

Every program looks for such marker files only IN backup directory.

But it's up to maintainers to decide what to do... Users will live with this, workaround or switch to another tool.

**Comment by rawtaz** [comment]
> Every program looks for such marker files only IN backup directory.

Are you saying that with these programs, if you ask them to back up `/foo/bar/star` when you have e.g. an `CACHEDIR.TAG` file in `/foo`, then they will back up/process the `/foo/bar/star` directory nonetheless? Did you actually try this to verify? I looked at their help text and it doesn't say anything about it.

**Comment by dionorgua** [comment]
Yes. I've tried every program in provided list with followed tree
```.
├── CACHEDIR.TAG
└── src
    ├── backup_me.txt
    ├── some_dir
    └── src2
        ├── CACHEDIR.TAG
        └── large_file

3 directories, 4 files
```
Every program backed up `backup_me.txt` file. The only difference I see is that some tools also saves `CACHEDIR.TAG` files and/or empty excluded directories. But every tool ignored `CACHEDIR.TAG` outside src dir.

**1. GNU tar**
```
dion@test:/tmp/sample_dir% tar czvf /tmp/a.tar.gz --exclude-caches $PWD/src
tar: Removing leading `/' from member names
/tmp/sample_dir/src/
/tmp/sample_dir/src/some_dir/
/tmp/sample_dir/src/src2/
tar: /tmp/sample_dir/src/src2/: contains a cache directory tag CACHEDIR.TAG; contents not dumped
/tmp/sample_dir/src/src2/CACHEDIR.TAG
/tmp/sample_dir/src/backup_me.txt

dion@test:/tmp/sample_dir% tar tzf /tmp/a.tar.gz                           
tmp/sample_dir/src/
tmp/sample_dir/src/some_dir/
tmp/sample_dir/src/src2/
tmp/sample_dir/src/src2/CACHEDIR.TAG
tmp/sample_dir/src/backup_me.txt
dion@test:/tmp/sample_dir% 
```

**2. borg**
```
dion@test:/tmp/sample_dir% borg create --exclude-caches /tmp/borg-repo::test $PWD/src 
Enter passphrase for key /tmp/borg-repo: 
dion@test:/tmp/sample_dir% borg list /tmp/borg-repo::test                             
Enter passphrase for key /tmp/borg-repo: 
drwxr-xr-x dion   dion          0 Thu, 2020-05-07 16:07:04 tmp/sample_dir/src
-rw-r--r-- dion   dion         30 Thu, 2020-05-07 15:48:57 tmp/sample_dir/src/backup_me.txt
drwxr-xr-x dion   dion          0 Thu, 2020-05-07 16:07:04 tmp/sample_dir/src/some_dir
```
**3. rclone**
```
dion@test:/tmp/sample_dir% rm -rf /tmp/rclone_out
dion@test:/tmp/sample_dir% rclone copy --exclude-if-present CACHEDIR.TAG $PWD/src /tmp/rclone_out
dion@test:/tmp/sample_dir% ls -l /tmp/rclone_out
total 4
-rw-r--r-- 1 dion dion 30 May  7 15:48 backup_me.txt
dion@test:/tmp/sample_dir% 
```

**4. duplicity**
```
dion@test:/tmp/sample_dir% duplicity --no-encryption --exclude-if-present CACHEDIR.TAG $PWD/src file:///tmp/duplicity_remote
Warning, found signatures but no corresponding backup files
Synchronizing remote metadata to local cache...
Deleting local /home/dion/.cache/duplicity/061738bba0247212f68dec51a62a01bc/duplicity-full-signatures.20200507T153013Z.sigtar.gz (not authoritative at backend).
Deleting local /home/dion/.cache/duplicity/061738bba0247212f68dec51a62a01bc/duplicity-full.20200507T153013Z.manifest (not authoritative at backend).
Last full backup date: none
No signatures found, switching to full backup.
--------------[ Backup Statistics ]--------------
StartTime 1588865426.96 (Thu May  7 18:30:26 2020)
EndTime 1588865426.96 (Thu May  7 18:30:26 2020)
ElapsedTime 0.00 (0.00 seconds)
SourceFiles 3
SourceFileSize 170 (170 bytes)
NewFiles 3
NewFileSize 170 (170 bytes)
DeletedFiles 0
ChangedFiles 0
ChangedFileSize 0 (0 bytes)
ChangedDeltaSize 0 (0 bytes)
DeltaEntries 3
RawDeltaSize 30 (30 bytes)
TotalDestinationSizeChange 204 (204 bytes)
Errors 0
-------------------------------------------------

dion@test:/tmp/sample_dir% duplicity list file:///tmp/duplicity_remote
Local and Remote metadata are synchronized, no sync needed.
Last full backup date: Thu May  7 18:30:26 2020
Thu May  7 16:07:04 2020 .
Thu May  7 15:48:57 2020 backup_me.txt
Thu May  7 16:07:04 2020 some_dir
dion@test:/tmp/sample_dir%
```
**5. rdiff_backup**
```
dion@test:/tmp/sample_dir% rdiff-backup --exclude-if-present CACHEDIR.TAG $PWD/src /tmp/rdiff_remote             
dion@test:/tmp/sample_dir% ls -l /tmp/rdiff_remote 
total 4
-rw-r--r-- 1 dion dion  30 May  7 15:48 backup_me.txt
drwx------ 3 dion dion 260 May  7 18:31 rdiff-backup-data
drwxr-xr-x 2 dion dion  40 May  7 16:07 some_dir
dion@test:/tmp/sample_dir% 
```

**Comment by rawtaz** [comment]
Given the inconsistency I'm not convinced that there's an established de facto standard here, but let's reopen it. Thanks for being thorough in that testing, it's very appreciated to get such a clear picture.

**Comment by dionorgua** [comment]
Thanks for reopening this!. 

There is no inconsistency in 'excluding' part (actually reason of this bug). All tools except `restic` do exactly same thing.

As about storing empty dir + `CACHEDIR.TAG` itself, it's a bit different. And I think that `restic` is doing right thing here (same as `tar`). It basically stores information that 'directory was present, but just excluded from backup'. Ideally `restic` should also store information about direct exclude flags somewhere in snapshot, but it's out of scope of this ticket.

**Comment by rawtaz** [comment]
There is inconsistence/no consensus on how to do the exclusion of the contents in the folder.

**Comment by MichaelEischer** [comment]
I think the current behavior is a bug: If you move the `src` folder from the sample dir into a new folder `src0` and then create a backup using `restic backup --exclude-if-present CACHEDIR.TAG src0/src`, then the content of the src folder is actually included in the backup.

**Comment by bohni** [comment]
So how about the ~/.cache directory? If this change would be implemented, I have to mark EVERY directory under ~/.cache with the CACHEDIR.TAG

And Broian ford, the author of the proposal of CACHEDIR.TAG, wrote (I marked the relevant part):

Application Semantics

The presence of a cache directory tag within a given directory indicates that the entire contents of that directory, **including any and all subdirectories underneath it**, consists of cached information that can be re-generated if necessary from appropriate source material located elsewhere.

See: https://bford.info/cachedir/

So I vote for closing this feature request (not bug)

**Comment by mafalb** [comment]
a tough one! here is my take

many applications have rules of precedence

1. implicit defaults
2. configuration file (overrides implicit defaults)
3. command line parameters (overrides configuration files)

If I request a backup for a directory *explicitly* via command line I would expect that it has more weight as a CACHEDIR.TAG (which I would see as an implicit default from above). Put it another way: I am the user and I know what I am doing.

Please consider to expand the issue to the following case:

```
/tmp/sample_dir% tree
.
├── CACHEDIR.TAG
└── src
    └── CACHEDIR.TAG 
    └── backup_me.txt
```

Otherwise there is no way to force a backup of a cachedir other than temporarily removing the CACHEDIR.TAG files.

**Comment by mafalb** [comment]
> And Brian Ford, the author of the proposal of CACHEDIR.TAG, wrote (I marked the relevant part):
> 
> Application Semantics
> 
> The presence of a cache directory tag within a given directory indicates that the entire contents of that directory, **including any and all subdirectories underneath it**, consists of cached information that can be re-generated if necessary from appropriate source material located elsewhere.
> 
> See: https://bford.info/cachedir/

He also wrote:

```
In short, to maintain robustness of 
backups in the face of security compromises, cache directory tags should only be treated as hints, never
as "the final word" on what should and should not be backed up.
```

How can I have "the final word" as it is implemented now?

**Comment by bohni** [comment]
Read the docs of restic

```
--exclude-caches                         excludes cache directories that are marked with a CACHEDIR.TAG file. See https://bford.info/cachedir/ for the Cache Directory Tagging Standard 
```



> Otherwise there is no way to force a backup of a cachedir other than temporarily removing the CACHEDIR.TAG files.
you don't have to

> How can I have "the final word" as it is implemented now?
use the right command line params (eg. don't use `--exclude-caches`)

**Comment by mafalb** [comment]
> > How can I have "the final word" as it is implemented now?
> 
> use the right command line params (eg. don't use `--exclude-caches`)

And it is not enabled by default. I missed that.

**Comment by MichaelEischer** [comment]
> So how about the ~/.cache directory? If this change would be implemented, I have to mark EVERY directory under ~/.cache with the CACHEDIR.TAG

Only if you run `backup ~/.cache/*` (and the CACHEDIR.TAG is in ~/cache). The point of this issue is to allow one to backup parts of a directory tree which contains a CACHEDIR.TAG in some parent folder, it restic is explicitly told to backup data within such a folder.
